plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -0,0 +1,618 @@
+} IspStatus;
+
+typedef enum {
+	ISP_CMD_ENTER_FW_UPDATE		= 0x01,
+	ISP_CMD_GET_PROJECT_ID_ADDR	= 0x02,
+	ISP_CMD_SYNC_IDENTIFY_CODE	= 0x03,
+	ISP_CMD_GET_FW_INFO		= 0x04,
+	ISP_CMD_FW_UPDATE_START		= 0x05,
+	ISP_CMD_FW_UPDATE_ISP_DONE	= 0x06,
+	ISP_CMD_FW_UPDATE_EXIT		= 0x07,
+	ISP_CMD_FW_UPDATE_RESET		= 0x08,
+} IspCmd;
+
+static void
+fu_rts54hub_rtd21xx_device_to_string (FuDevice *module, guint idt, GString *str)
+{
+	FuRts54hubRtd21xxDevice *self = FU_RTS54HUB_RTD21XX_DEVICE (module);
+	fu_common_string_append_kx (str, idt, "TargetAddr", self->target_addr);
+	fu_common_string_append_kx (str, idt, "I2cSpeed", self->i2c_speed);
+	fu_common_string_append_kx (str, idt, "RegisterAddrLen", self->register_addr_len);
+}
+
+static FuRts54HubDevice *
+fu_rts54hub_rtd21xx_device_get_parent (FuRts54hubRtd21xxDevice *self, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent (FU_DEVICE (self));
+	if (parent == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "no parent set");
+		return NULL;
+	}
+	return FU_RTS54HUB_DEVICE (parent);
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_set_quirk_kv (FuDevice *device,
+					 const gchar *key,
+					 const gchar *value,
+					 GError **error)
+{
+	FuRts54hubRtd21xxDevice *self = FU_RTS54HUB_RTD21XX_DEVICE (device);
+	/* load target address from quirks */
+	if (g_strcmp0 (key, "Rts54TargetAddr") == 0) {
+		guint64 tmp = fu_common_strtoull (value);
+		if (tmp <= 0xff) {
+			self->target_addr = tmp;
+			return TRUE;
+		}
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid target address");
+		return FALSE;
+	}
+
+	/* load i2c speed from quirks */
+	if (g_strcmp0 (key, "Rts54I2cSpeed") == 0) {
+		guint64 tmp = fu_common_strtoull (value);
+		if (tmp < FU_RTS54HUB_I2C_SPEED_LAST) {
+			self->i2c_speed = tmp;
+			return TRUE;
+		}
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid I²C speed");
+		return FALSE;
+	}
+
+	/* load register address length from quirks */
+	if (g_strcmp0 (key, "Rts54RegisterAddrLen") == 0) {
+		guint64 tmp = fu_common_strtoull (value);
+		if (tmp <= 0xff) {
+			self->register_addr_len = tmp;
+			return TRUE;
+		}
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid register address length");
+		return FALSE;
+	}
+
+	/* failed */
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "quirk key not supported");
+	return FALSE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_i2c_write (FuRts54hubRtd21xxDevice *self,
+				      guint8 target_addr, guint8 sub_addr,
+				      const guint8 *data, gsize datasz,
+				      GError **error)
+{
+	FuRts54HubDevice *parent;
+
+	parent = fu_rts54hub_rtd21xx_device_get_parent (self, error);
+	if (parent == NULL)
+		return FALSE;
+	if (!fu_rts54hub_device_vendor_cmd (parent, 1, error))
```

Comment:
magic vendor command should probably have a define

Reply:
@AnyProblem what's 0x01 mean?

Reply:
0x01 means general vendor cmd enable, 0x02 access flash right, 0x0 disable

Reply:
I've added this as a follow up commit.